text/chapter10.md

```diff
@@ -391,416 +437,935 @@ foreign import isUndefined :: forall a. Undefined a -> Boolean
 
-> runExcept (decodeJSON "true" :: F Boolean)
-Right true
+```hs
+foreign import alert :: String -> Effect Unit
+```
 
-> runExcept (decodeJSON "[1, 2, 3]" :: F (Array Int))
-Right [1, 2, 3]
+```js
+exports.alert = msg => () =>
+  window.alert(msg);
 ```
 
-Recall that in the `Either` monad, the `Right` data constructor indicates success. Note however, that invalid JSON, or an incorrect type leads to an error:
+We want this alert to appear when either:
+- A user attempts to save a form with validation errors.
+- The state cannot be retrieved from local storage.
 
-```text
-> runExcept (decodeJSON "[1, 2, true]" :: F (Array Int))
-(Left (NonEmptyList (NonEmpty (ErrorAtIndex 2 (TypeMismatch "Int" "Boolean")) Nil)))
+That is accomplished by simply replacing `log` with `alert` on these lines:
+```hs
+Left errs -> alert $ "There are " <> show (length errs) <> " validation errors."
+
+alert $ "Error: " <> err <> ". Loading examplePerson"
 ```
 
-The `foreign-generic` library tells us where in the JSON document the type error occurred.
+ ## Exercises
 
-## Handling Null and Undefined Values
+ 1. (Easy) Write a wrapper for the `removeItem` method on the `localStorage` object, and add your foreign function to the `Effect.Storage` module.
+ 1. (Medium) Add a "Reset" button that, when clicked, calls the newly-created `removeItem` function to delete the "person" entry from local storage.
+ 1. (Easy) Write a wrapper for the `confirm` method on the JavaScript `Window` object, and add your foreign function to the `Effect.Alert` module.
+ 1. (Medium) Call this `confirm` function when a users clicks the "Reset" button to ask if they're sure they want to reset their address book.
 
-Real-world JSON documents contain null and undefined values, so we need to be able to handle those too. `foreign-generic` solves this problem with the 'Maybe' type constructor to represent missing values.
+## Conclusion
 
-```text
-> import Prelude
-> import Foreign.NullOrUndefined
+In this chapter, we've learned how to work with foreign JavaScript code from PureScript and we've seen the issues involved with writing trustworthy code using the FFI:
+- We've seen the importance of ensuring that foreign functions have correct representations.
+- We learned how to deal with corner cases like null values and other types of JavaScript data, by using foreign types, or the `Json` data type.
+- We saw how to safely serialize and deserialize JSON data.
+
+For more examples, the `purescript`, `purescript-contrib` and `purescript-node` GitHub organizations provide plenty of examples of libraries which use the FFI. In the remaining chapters, we will see some of these libraries put to use to solve real-world problems in a type-safe way.
+
+# Addendum
```

Comment:
In the spirit of being an "Introductory Text", I think the most important content to cover in this chapter is how to call JavaScript from PureScript, as is required for building the Address Book app. The details on runtime representation and calling PS from JS don't seem as necessary. I think it would be best to let these sections serve as an addendum, or even just live in the docs repo where they are partially duplicated already.

Some related discussion:
https://functionalprogramming.slack.com/archives/C04NA444H/p1588600540352500
https://functionalprogramming.slack.com/archives/C04NA444H/p1588600910356600
https://functionalprogramming.slack.com/archives/C04NA444H/p1588610399361200

Summary:
Phil originally had a JS app with a pure PS core when he wrote the book, but that use case is less common today, and creating a "reverse ffi" module for effectful purescript seems outside the scope of this book.

This guide has some details on embedding effectful PS within a React JS app.
https://thomashoneyman.com/articles/replace-react-components-with-purescript/#replacing-a-react-component-with-purescript-react

The simpler argonaut workflow replaced these sections in the original text:
- Working With Untyped Data
- Handling Null and Undefined Values
- Generic JSON Serialization